### ISSUE TITLE
Introduce rule that reports unused internal members

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#964]
As far as I'm aware, there's no analyzer that reports a diagnostic for unused internal members.
There's [IDE0051](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0051), but it only reports unused **private** members.

It would be nice to have a rule that reports:
* unused members in internal (top-level or nested) classes (regardless of their accessibility)
* unused internal members in (top-level or nested) public classes

This rule should of course not fire for the **Main** method in the entry point of an executable application.

Perhaps there's a good reason why no analyzer implements such a rule.

PS. Don't hesitate to tell me if I'm pushing it here

**Comment by meziantou** [comment]
I think that's how the latest rule (MA0182 - Avoid **unused internal types**) works. That's also mean it differ from CA1812 which check for "uninstantiated" classes.

**Comment by drieseng** [comment]
MA0182 does not report unused internal members in a public class.
MA0182 does not report unused non-private members in an internal class (unused private members are already reported by **IDE0051**).

**Comment by meziantou** [comment]
I read types instead of members... Then, it would be a new rule.
However, there are more chance to get false-positive because of `InternalsVisibleTo`. I think I need to take some time to think about it.

> MA0182 does not report unused internal members in a public class.

Looks like you already opened an issue for this one #965

**Comment by drieseng** [comment]
> I read types instead of members... Then, it would be a new rule. However, there are more chance to get false-positive because of `InternalsVisibleTo`. I think I need to take some time to think about it.

I don't see why, but take your time. Any members that are only ever accessed in unit test projects can be consider unused.

> > MA0182 does not report unused internal members in a public class.
> 
> Looks like you already opened an issue for this one [#965](https://github.com/meziantou/Meziantou.Analyzer/issues/965)

No, that one had nothing to do with the current "issue".

**Comment by drieseng** [comment]
I think the following should never be reported as unused:
* abstract member
* member that overrides a member of a base class.
* member that implements an abstract member of a base class.
* member that implement an interface
* virtual member?
* ...

**Comment by drieseng** [comment]
@meziantou: there has been a [request](https://github.com/dotnet/roslyn/issues/68029) to update **IDE0051** to report unused internal members, but it's not moving anywhere.

**Comment by meziantou** [comment]
It's a bit more complicated than your list
- `ModuleInitializerAttribute`
- `DynamicallyAccessedMembersAttribute` (and use the ctor value)
- Assembly Entrypoint (can be private but should be considered used)
- What about interface/virtual/abstract members that are not used (maybe implemented but no actual usage)

> abstract member

Why? If there is no usage, then I think it should be considered as unused

> virtual member?

Same

**Comment by drieseng** [comment]
> > abstract member
> 
> Why? If there is no usage, then I think it should be considered as unused

The abstract member could never be invoked from the class that defines the member, but invoked in or on a deriving class.

> > virtual member?
> 
> Same

The virtual member on the class that defines the member may never be used, but overridden in a derived class (and that one may effectively be used).

**Comment by meziantou** [comment]
I made a quick attempt to create the rules and here're some implementation notes:
- Entrypoint should not be reported
- Methods decorated with `[ModuleInitializerAttribute]` should not be reported
- For types decarated with `[DynamicallyAccessMembers]`, read the enum value and report members as used accordingly
- `ComImport` / `ComVisible` / `CoClass` class members should not be reported
- `[UnmanagedCallersOnly]` methods should not be reported
- `[DebuggerDisplay("{member}")]` should be parsed and find members. Note that expression can contains nested members (e.g. `Prop1.Value.Method(Prop2).Result`), so the parsing is best effort. (note: there is already another rule that parse the DebuggerDisplay value, so it can be used as a reference for this rule).
- Ignore unit test methods (MStest, xunit, nunit, tunit attributes)
- Static constructors should not be reported
- Finalizers should not be reported
- Compiler generated members should not be reported (for instance, baking fields for auto properties)
- partial methods without implementation should not be reported (mostly used by generated code)
- `nameof(member)` should not be considered as a usage
- Properties should differentiate getter and setter
- Best effort for reflection (for instance, typeof(T).GetMethod("name"))? `[DynamicallyAccessMembers]` should be the way to go.
- Collection expression should mark the Create method as used. ICollectionOperation seems to provide the Create symbol.
- `foreach` should mark the `GetEnumerator` and its `Current` and `MoveNext` methods as used. Note that `GetEnumerator` can be an extension method. IForEachLoopOperation doesn't provide the GetEnumerator symbol.
- `Deconstruct` methods should be detected from usage `var (x, y) = value;`. Note that is can be an etension method.
- `await` should mark `GetAwaiter`, `GetResult`, `IsCompleted` as used. Note that `GetAwaiter` can be an extension method. IAwaitOperation doesn't provide the GetAwaiter symbol.
- `await using` statement should mark the `DisposeAsync` method as used even if the class doesn't implement `AsyncIDisposable` (not required by the language)
- Copy constructors should be mark as used when using `with` (record and struct)
- Overriding a member from an internal class/interface is not a usage of the base member. Using the member directly is a usage. If the base member is used, then all inheritors are considered as used.
- Operators can be ignored as they must be public. So, there is no need to analyze implicit or explicit conversions.
- Cref are not considered as usage
- Type decorated with StructLayout Sequential should not be reported, as removing a member may break layout.

Roslyn doesn't provide the binding information for all pattern-based usages (foreach, await, etc.), this means the logic should be replicated and can be tricky because some of them allow extension methods.

Open questions: 
- How to handle `InternalsVisibleTo`?
- Are there other patterns to detect?
- Ignore serialization properties (XML, YAML, JSON)?
- Should it report fields that are never read?

Based on that investigation, this seems to be a large task. Also, there are lots of potential bugs in the implementation. If the rule is implemented, it should be disabled by default.

Complex patterns to detect

```
// DynamicallyAccessedMembers
public static object? CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type type, params object?[]? args)

public class OptionsManager<[DynamicallyAccessedMembers(Options.DynamicallyAccessedMembers)] TOptions>

[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicNestedTypes | DynamicallyAccessedMemberTypes.NonPublicNestedTypes)]
public virtual TypeInfo? GetDeclaredNestedType(string name) => throw null;

[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
private readonly Type _type;

[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)]
[return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)]
public override Type? GetInterface(string name, bool ignoreCase)

public Type? this[string key]
{
    [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
    get => throw null;
    [param: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]
    set => throw null;
}
```

**Comment by drieseng** [comment]
> I made a quick attempt to create the rules and here're some implementation notes:
> 
> * Ignore unit test methods (MStest, xunit, nunit, tunit attributes)

Don't these always have to be public anyway. For MSTest I'm pretty sure (but I can double check later).

> * partial methods should not be reported

why not ?

> Open questions:
> 
> * How to handle `InternalsVisibleTo`?

I think MA0182 and this rule should be align. I'm ok with matching MA0182, or if necessary have an option to either opt out or opt in in the precense of `InternalsVisibleTo`. There should definitely be a way to report unused types and members even if the internals are exposed to other assemblies.

> * Are there other patterns to detect?

Not that I can think of right now. Regardless of your extensive list above, I'm sure we'll discover more cases w

> * Ignore serialization properties (XML, YAML, JSON)?

Can you elaborate on this one? Do you mean properties that are annotated with serialization attributes?

> Based on that investigation, this seems to be a large task.

Is Copilot up for (at least part of) a task like this?

> Also, there are lots of potential bugs in the implementation. If the rule is implemented, it should be disabled by default.

Perhaps initially document it as being "experimental".

**Comment by meziantou** [comment]
> Can you elaborate on this one? Do you mean properties that are annotated with serialization attributes?

Yes, properties having serialization attributes such as

```c#
internal class Sample
{
    [JsonPropertyName("foo")]
    public class Foo { get; set; }
}
```

In the case of deserialization only, the setter may never be used in the code.

**Comment by meziantou** [comment]
Just to let you know, I won't work on this issue. There are too many cases to support to reduce the number of false-positives while being relevant. Also, Roslyn doesn't provide everything out of the box, so you have to deal with IOperation (easy) and Syntax/SemanticModel (more fragile). For instance, method group is not correctly exposed through IOperation.
Also, the hard part is to implement handle `DynamicallyAccessMembers`. Basically, this means duplicating the non-trivial ILLink analyzer (https://github.com/dotnet/runtime/tree/main/src/tools/illink/src/ILLink.RoslynAnalyzer). Also, if you look at the code, there are chances that it requires maintenance at each new C# release.
Last but not least, when writing the tests, I'm not always sure what's the expected behavior.

So, the decision is mainly motivated by the initial complexity and also the maintenance cost at each new C# feature.

Note: The initial (broken) implementation is available here: https://github.com/meziantou/Meziantou.Analyzer/commit/8dbff37fa3812808ac3d58fb223171bd50d006cd#diff-fcdb7281d565ce2bf019e77304df4940e34e8fe70e145d6a36399abac94b130b
